src/legacy/timer/timer.go

```diff
@@ -0,0 +1,206 @@
+package timer
+
+import (
+	"unsafe"
+)
+
+type Timer struct {
+	Flags        uint32 // 0, bit 1: raw(set) or interp(unset), bit 2: set when updated
+	Current      uint32 // 4, current value
+	Target       uint32 // 8, target value
+	DeltaPerTick uint32 // 12, (interp only) the amount to add per tick
+	MaxTickDelta uint64 // 16, (interp only) the maximum delta per tick
+	LastUpdated  uint64 // 24, the last tick the value was updated
+}
+
+var _ = [1]struct{}{}[32-unsafe.Sizeof(Timer{})]
+
+func (timer *Timer) C() unsafe.Pointer {
+	return unsafe.Pointer(timer)
+}
+
+type TimerGroup struct {
+	// Timers[0]: usually internal value, ranged from 0 to 0x4000
+	// Timers[1]: For UI display? ranged from 0 to 100
+	// Timers[2]: Not quite sure. Ranged from 0 to 0x2000
+	Timers [3]Timer
+}
+
+var _ = [1]struct{}{}[96-unsafe.Sizeof(TimerGroup{})]
+
+func (timerGroup *TimerGroup) C() unsafe.Pointer {
+	return unsafe.Pointer(timerGroup)
+}
+
+func TimerInit(self *Timer, target int32) int32 {
+	self.Flags = 0
+	self.LastUpdated = uint64(PlatformTicks())
+	TimerSetParams(self, 0x3E8, 0x4000)
+	TimerSetRaw(self, uint32(target))
+	return TimerUpdate(self)
+}
+
+func Sub_4862E0(a3 unsafe.Pointer, a4 int32) int32 {
+	return TimerInit((*Timer)(a3), a4)
+}
+
+func TimerSetParams(self *Timer, max_delta uint32, interp_velocity uint32) {
+	self.MaxTickDelta = uint64(max_delta)
+	self.DeltaPerTick = (uint32(interp_velocity) << 16) / max_delta
+}
+
+func Sub_486380(a1 unsafe.Pointer, a2 uint32, a3 int32, a4 uint32) int32 {
+	TimerSetParams((*Timer)(a1), a2, a4)
+	return 0
+}
+
+func TimerSetRaw(self *Timer, target uint32) {
+	var a1 unsafe.Pointer = self.C()
+	*(*uint32)(a1) |= 1
+	*(*uint32)(unsafe.Add(a1, 4*2)) = uint32(target) << 16
+}
+
+func Sub_486320(a1 unsafe.Pointer, a2 uint32) {
+	TimerSetRaw((*Timer)(a1), a2)
+}
+
+func TimerSetInterp(self *Timer, target uint32) int32 {
+	var a1 unsafe.Pointer = self.C()
+	var (
+		v2 int64
+		v3 int32
+	)
+	*(*uint32)(unsafe.Add(unsafe.Pointer(&v2), 4*0)) = *(*uint32)(unsafe.Add(a1, 8))
+	v3 = int32(*(*uint32)(unsafe.Add(a1, 4)))
+	*(*uint32)(a1) &= 0xFFFFFFFE
+	if uint32(int32(v2)) == uint32(v3) {
+		v2 = int64(PlatformTicks())
+		*(*uint64)(unsafe.Add(a1, 24)) = uint64(v2)
+	}
```

Comment:
Notice how the code takes address from a local variable `v2`. This is dangerous! Go stack behaves differently to C stack. Go may allocate variables on the heap and then destroy them, if the pointer suddenly disappears (e.g. is cast to `int` type).

I would suggest rewriting this part. I noticed that first 4 bytes of `v2` are used for `if`, and the all 8 bytes are overwritten in the `if` body. Thus, these are two independent variables:
```suggestion
	v2 := *(*uint32)(unsafe.Add(a1, 8))
	v3 := *(*uint32)(unsafe.Add(a1, 4))
	*(*uint32)(a1) &= 0xFFFFFFFE
	if v2 == v3 {
		*(*uint64)(unsafe.Add(a1, 24)) = PlatformTicks()
	}
```

Since it already fixes most of the function body, you could easily fix the rest of unsafe in this function too.

Reply:
Simply missed during migration (haha;;)